packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -511,3 +461,40 @@ export class PendingStateManager implements IDisposable {
         this.stateHandler.setFlushMode(savedFlushMode);
     }
 }
+
+/**
+ * Helper class that manages partial batch & rollback.
+ */
+export class BatchManager {
+    private pendingBatch: BatchMessage [] = [];
+
+    public push(message: BatchMessage) {
+            this.pendingBatch.push(message);
+    }
+
+    public get empty() { return this.pendingBatch.length === 0; }
+
+    public popBatch() {
+        const batch = this.pendingBatch;
+        this.pendingBatch = [];
+        return batch;
+    }
+
+    /**
+     * Capture the pending state at this point
+     */
+     public checkpoint() {
+        const startPoint = this.pendingBatch.length;
+        return {
+            rollback: (process: (message: BatchMessage) => void) => {
+                for (let i = this.pendingBatch.length; i > startPoint;) {
+                    i--;
```

Comment:
#nit why not `i--` in the `for` statement?

Reply:
3 more bytes to add :)

for (let i = this.pendingBatch.length - 1; i >= startPoint; i--;) {

but I can go either way :)